core/lib/multivm/src/versions/vm_latest/tracers/prestate_tracer.rs

```diff
@@ -0,0 +1,235 @@
+            let cloned_storage = Rc::clone(&storage);
+            let mut initial_storage_ref = cloned_storage.as_ref().borrow_mut();
+            let keys = initial_storage_ref
+                .modified_storage_keys()
+                .keys()
+                .cloned()
+                .collect::<Vec<_>>();
+
+            let res = keys
+                .iter()
+                .map(|k| {
+                    (
+                        *(k.account().address()),
+                        Account {
+                            balance: Some(h256_to_u256(
+                                initial_storage_ref.read_value(&get_balance_key(k.account())),
+                            )),
+                            code: Some(h256_to_u256(
+                                initial_storage_ref
+                                    .read_value(&get_code_key(k.account().address())),
+                            )),
+                            nonce: Some(h256_to_u256(
+                                initial_storage_ref
+                                    .read_value(&get_nonce_key(k.account().address())),
+                            )),
+                            storage: Some(get_storage_if_present(
+                                k.account(),
+                                initial_storage_ref.modified_storage_keys(),
+                            )),
+                        },
+                    )
+                })
+                .collect::<State>();
+            self.pre = res;
+        }
+    }
+}
+
+impl<S: WriteStorage, H: HistoryMode> VmTracer<S, H> for PrestateTracer {
+    fn initialize_tracer(&mut self, _state: &mut ZkSyncVmState<S, H>) {}
+
+    fn finish_cycle(
+        &mut self,
+        _state: &mut ZkSyncVmState<S, H>,
+        _bootloader_state: &mut BootloaderState,
+    ) -> TracerExecutionStatus {
+        TracerExecutionStatus::Continue
+    }
+
+    fn after_vm_execution(
+        &mut self,
+        state: &mut ZkSyncVmState<S, H>,
+        _bootloader_state: &BootloaderState,
+        _stop_reason: crate::interface::tracer::VmExecutionStopReason,
+    ) {
+        let modified_storage_keys = state.storage.storage.inner().get_modified_storage_keys();
+        if self.config.diff_mode {
+            let diff = modified_storage_keys
+                .clone()
+                .keys()
+                .copied()
+                .collect::<Vec<_>>();
+
+            let res = diff
+                .iter()
+                .map(|k| get_account_data(k, state, &modified_storage_keys))
+                .collect::<State>();
+            self.post = res;
+        } else {
+            let read_keys = &state.storage.read_keys;
+            let map = read_keys.inner().clone();
+            let storage_keys_read = map.keys().copied().collect::<Vec<_>>();
+            let res = storage_keys_read
+                .iter()
+                .map(|k| get_account_data(k, state, &modified_storage_keys))
+                .collect::<State>();
```

Comment:
nit: 
You don't need an intermediate representation of storage_keys_read. 
You can start iterating over keys

p.s. same above